pkg/provider/bao/bao.go

```diff
@@ -0,0 +1,174 @@
+// Copyright © 2023 Bank-Vaults Maintainers
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bao
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"os"
+	"os/signal"
+	"strings"
+
+	injector "github.com/bank-vaults/internal/pkg/baoinjector"
+	bao "github.com/bank-vaults/vault-sdk/vault"
+
+	"github.com/bank-vaults/secret-init/pkg/common"
+	"github.com/bank-vaults/secret-init/pkg/provider"
+)
+
+const (
+	ProviderName = "bao"
+	Re           = `(bao:)(.*)#(.*)`
+)
+
+type Provider struct {
+	isLogin        bool
+	client         *bao.Client
+	injectorConfig injector.Config
+	secretRenewer  injector.SecretRenewer
+	fromPath       string
+	revokeToken    bool
+}
+
+type sanitized struct {
+	secrets []provider.Secret
+	login   bool
+}
+
+// BAO_* variables are not populated into this list if this is not a login scenario.
+func (s *sanitized) append(key string, value string) {
+	envType, ok := sanitizeEnvmap[key]
+	// If the key being appended is not present in sanitizeEnvmap, it signifies that
+	// it is not a BAO_* variable.
+	// Additionally, in a login scenario, we include BAO_* variables in the secrets list.
+	if !ok || (s.login && envType.login) {
+		// Path here is actually the secret's key,
+		// An example of this can be found at the LoadSecrets() function below
+		secret := provider.Secret{
+			Path:  key,
+			Value: value,
+		}
+
+		s.secrets = append(s.secrets, secret)
+	}
+}
+
+func NewProvider(config *Config) (*Provider, error) {
```

Comment:
lets pass the app config directly here along with the actual provider config to avoid redoing it internally, something like:
```suggestion
func NewProvider(config *Config, appConfig *common.Config) (*Provider, error) {
```